### PR TITLE
Adding pricing/fiat conversion functionality

### PR DESCRIPTION
**Pull request by gene-telligent**
As described in https://github.com/monero-project/monero-gui/issues/1484.

A few things of note:

## Connections and security
This module uses the [QNetworkAccessManager](http://doc.qt.io/qt-5/qnetworkaccessmanager.html) to perform HTTP requests to the user's choice of a centralized server. As such, it introduces an insecure element for the benefit of convenience. This feature is off by default, and builds have to explicitly include the WITH_PRICES config option. As part of this pull request, I've added the WITH_PRICES flag to the release and debug builds in build.sh, but that can be easily reverted if desired (switching from an opt-out to an opt-in).

## Pricing providers
Out of the box I include support for CMC, Binance, and CryptoCompare. Other providers can be added by subclassing the PriceSource class.

## Currencies
Right now the choices are fairly arbitrary (I don't think I left out any major ones), but new currencies can be added very easily by creating new Currency objects.

I'm extremely open to feedback here; C++ isn't my first language so I'm sure there will be some issues or updates.

### CODE REVIEW

**Comment by selsta** [comment]
I think it would be good to explain what the left (source) and the right (currency) dropdown menus are, as they are both blank when opening the first time.

screenshot 2018-11-13 at 20 11 55

**Comment by skftn** [comment]
Cool contribution @gene-telligent. Ill review it, but it will take some time.

**Comment by cryptochangements34** [comment]
By pinging these services for the exact amount in your wallet you're basically announcing how much money you have, isnt that bad for privacy?

**Comment by adamgall** [comment]
> By pinging these services for the exact amount in your wallet you're basically announcing how much money you have, isnt that bad for privacy?

I haven't yet read through the PR, but why would the wallet need to tell any third party service how much XMR it holds? All it theoretically needs to do is get the current exchange rate, then (locally) multiply that by the number of XMR in the wallet.

**Comment by cryptochangements34** [comment]
That's my bad. Nevermind 😱

**Comment by lewisisgood** [comment]
Hey @xmrdsc! Any update on when you may be able to look at this PR? Not trying to hustle you, just trying to get a sense of the timeline for this awesome feature! Thanks :)

**Comment by xmrdsc** [comment]
@lewisisgood Hey, I'm positive this will make it into the next release, assuming PR passes the code review. Also will have to see what others think but largely I think consensus is there.

File: src/prices/PriceManager.cpp [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+#include "Price.h"
+#include "Currency.h"
+#include "PriceSource.h"
+#include "CurrencySelectorModel.h"
+#include "PriceSourceSelectorModel.h"
+
+
+namespace {
+    static const int DEFAULT_REFRESH_PERIOD_MILLISECONDS = 15 * 1000;
+    static const QByteArray USER_AGENT = "monero-gui/0.12.0 (Qt)";
```
Comment (xmrdsc) [inline src/prices/PriceManager.cpp]: Best not to advertise this, for privacy reasons. Use something standard like `Mozilla 5.0`.

**Comment by xmrdsc** [comment]
@lewisisgood Update regarding the merging of this PR; it will most likely make 14.1, which will be out in a few weeks. In the meantime, we'll need a rebase. If @gene-telligent can't find the time, I can rebase, but would rather not ;)

**Comment by gene-telligent** [comment]
@xmrdsc Rebased! Sorry about the delay.

File: src/prices/Currency.cpp [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+int Currency::precision() const
+{
+    return m_precision;
+}
+
+QString Currency::format(qreal amount) const
+{
+    return QLocale().toCurrencyString(amount, m_symbol, m_precision);
```
Comment (selsta) [inline src/prices/Currency.cpp]:
```suggestion
    return QLocale(QLocale().language(), QLocale().country()).toCurrencyString(amount, m_symbol, m_precision);
```

See here: https://forum.qt.io/topic/88146/qlocale-tocurrencystring-ignores-precision/3

**Comment by selsta** [comment]
```
diff --git a/pages/settings/SettingsLayout.qml b/pages/settings/SettingsLayout.qml
index 08afdf6..4c4b9e5 100644
--- a/pages/settings/SettingsLayout.qml
+++ b/pages/settings/SettingsLayout.qml
@@ -164,24 +164,6 @@ Rectangle {
             }
         }
 
-        MoneroComponents.StandardButton {
-            visible: !persistentSettings.customDecorations
-            Layout.topMargin: 10 * scaleRatio
-            small: true
-            text: qsTr("Change language") + translationManager.emptyString
-
-            onClicked: {
-                appWindow.toggleLanguageView();
-            }
-        }
-
-        MoneroComponents.TextBlock {
-            visible: isMobile
-            font.pixelSize: 14
-            textFormat: Text.RichText
-            Layout.fillWidth: true
-            text: qsTr("No Layout options exist yet in mobile mode.") + translationManager.emptyString;
-        }
         //! Manage pricing
 
         RowLayout {
@@ -261,6 +243,25 @@ Rectangle {
 
             z: parent.z + 1
         }
+
+        MoneroComponents.StandardButton {
+            visible: !persistentSettings.customDecorations
+            Layout.topMargin: 10 * scaleRatio
+            small: true
+            text: qsTr("Change language") + translationManager.emptyString
+
+            onClicked: {
+                appWindow.toggleLanguageView();
+            }
+        }
+
+        MoneroComponents.TextBlock {
+            visible: isMobile
+            font.pixelSize: 14
+            textFormat: Text.RichText
+            Layout.fillWidth: true
+            text: qsTr("No Layout options exist yet in mobile mode.") + translationManager.emptyString;
+        }
     }
 
     Component.onCompleted: {
```

This order makes more sense.

**Comment by selsta** [comment]
```
diff --git a/pages/settings/SettingsLayout.qml b/pages/settings/SettingsLayout.qml
index 08afdf6..863174a 100644
--- a/pages/settings/SettingsLayout.qml
+++ b/pages/settings/SettingsLayout.qml
@@ -264,6 +264,10 @@ Rectangle {
     }
 
     Component.onCompleted: {
+        priceSourceDropDown.currentIndex = appWindow.persistentSettings.currencyConversionSourceIndex;
+        priceSourceDropDown.update();
+        currencyDropDown.currentIndex = appWindow.persistentSettings.currencyConversionCurrencyIndex;
+        currencyDropDown.update();
         console.log('SettingsLayout loaded');
     }
 }
```

Fixes a bug with column text not showing.

**Comment by selsta** [comment]
@gene-telligent Thank you for your contribution! I’ve opened #2051 so that we can apply changes on top of this PR.

+resolved

**Pull request closed without merging**